### DISCUSSION TITLE
Add support for a node-lambda versioning

### DISCUSSION THREAD

**Discussion by Garethp** [renovatebot/renovate#32313] [Suggest an Idea]
### What would you like Renovate to be able to do?

I'd like to be able to update nodejs version with the schedule of AWS Lambdas. The current `node` versioning module waits for that node version to be released as LTS however Lambdas seem to lag the LTS versions by up to a month. I'd like to be able to keep my `.nvmrc` and docker image in-step with my Lambda versions and I think the correct method would be to create a new `node-lambda` versioning option that reflects AWS Lambda

### If you have any ideas on how this should be implemented, please tell us here.

The list of current Lambda versions and their EOL date can be found in JSON format at [https://endoflife.date/api/aws-lambda.json](https://endoflife.date/api/aws-lambda.json). I think this could be implemented by

* [ ] Generating static data based off of `endoflife.date` for `node-lambda`
* [ ] Adding a `lambda-node` versioning module in `lib/modules/versioning`

I'd be happy to try contributing this myself if this is a feature Renovate would be happy to accept

**Comment by rarkins** [comment]
Renovate already has an endoflife datasource, so you should check in case you can use that. You likely need to combine that with a custom regex manager and not write any new versioning either.

**Reply by Garethp** [reply]
That's what I'm doing to upgrade the instances of Lambda definitions in my code, where I'm struggling is keeping additional runtime dependencies such as the `.nvmrc` and `node` docker images in Gitlab CI. I agree a custom regex manager could likely solve those as well, but it seems like a pretty poor solution to write custom managers for things that Renovate is already picking up just to handle the scheduling differences of versioning. Considering that I'm hardly the only one looking at Lambda upgrading (and likely also at keeping other dependencies in line), I thought that contributing a versioning module for Lambdas might be a better use of my time and effort

**Reply by viceice** [reply]
try override datasource 

https://docs.renovatebot.com/configuration-options/#overridedatasource

**Reply by rarkins** [reply]
I think I now understand what you mean and agree with it. This type of thing could/should be done using either versioning or a datasource. One important thing though is that we don't want versioning to need to do network calls. The node versioning module is an example of using a semi-automatic, semi-static approach: https://github.com/renovatebot/renovate/tree/main/lib/modules/versioning/node

**Reply by Garethp** [reply]
That's what I was thinking. My implementation proposal would involve adding a `tools/static-data/generate-lambda-node-schedule.mjs` file and a `update-static-data:lambda-node-schedule` in the `package.json` so that the static data would be generated at build time rather than runtime.

If you'd be happy accepting this as a feature, I'll happily put it together and throw over an MR for it

**Reply by Garethp** [reply]
Sorry for the delay, I ended up trying a couple of different implementations before settling on the simplest one. I've opened an MR at #32480
